File: utils/badiDate.js
// Badi calendar constants
const BADI_EPOCH = new Date(Date.UTC(1844, 2, 21)); // March 21, 1844
// English-transliterated Badi month names (ASCII, no diacritics)
// English translations of the 20 Badi month names (in calendar order)
const BADI_MONTHS = [
  'Splendor',       // Bahá
  'Glory',          // Jalál
  'Beauty',         // Jamál
  'Grandeur',       // ‘Aẓamat
  'Light',          // Núr
  'Mercy',          // Raḥmat
  'Words',          // Kalimát
  'Perfection',     // Kamál
  'Names',          // Asmá’
  'Might',          // ‘Izzat
  'Will',           // Mashíyyat
  'Knowledge',      // ‘Ilm
  'Power',          // Qudrat
  'Speech',         // Qawl
  'Questions',      // Masá’il
  'Honour',         // S͟haraf
  'Sovereignty',    // Sulṭán
  'Dominion',       // Mulk
  'The Days of Há', // Ayyám-i-Há
  'Loftiness',      // ‘Alá’
];

// Helper: check if Gregorian year is a leap year
function isGregorianLeapYear(year) {
  return (year % 4 === 0 && year % 100 !== 0) || (year % 400 === 0);
}

// Get Naw-Rúz (March 21) of the given Gregorian year
function getNawRuzDate(gregorianYear) {
  return new Date(Date.UTC(gregorianYear, 2, 21)); // March 21
}

// Exported function: pass in a JS Date object, get Badi date
function getBadiDate(date) {
  const utcDate = new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()));

  //not the most acurate, this could be fixed.
  let badiYear = date.getFullYear() - 1843;
  let nawRuzThisYear = getNawRuzDate(date.getFullYear());

  if (utcDate < nawRuzThisYear) {
    badiYear -= 1;
    nawRuzThisYear = getNawRuzDate(date.getFullYear() - 1);
  }

  const daysSinceNawRuz = Math.floor((utcDate - nawRuzThisYear) / (1000 * 60 * 60 * 24));
  let dayOfYear = daysSinceNawRuz + 1;

  let badiMonth, badiDay;

  const isLeap = isGregorianLeapYear(date.getFullYear());

  if (dayOfYear <= 18 * 19) {
    badiMonth = Math.ceil(dayOfYear / 19);
    badiDay = ((dayOfYear - 1) % 19) + 1;
  } else if (dayOfYear <= 18 * 19 + (isLeap ? 5 : 4)) {
    badiMonth = 20; // Ayyám-i-Há
    badiDay = dayOfYear - (18 * 19);
  } else {
    badiMonth = 19; // ‘Alá’
    badiDay = dayOfYear - (18 * 19 + (isLeap ? 5 : 4));
  }

  return {
    badiDay,
    badiMonthName: BADI_MONTHS[badiMonth - 1],
    badiYear,
    formatted: `${badiDay} ${BADI_MONTHS[badiMonth - 1]} ${badiYear} BE`,
  };
}

module.exports = { getBadiDate };

File: ios/LiquidSpiritCalendarWidget/BadiDate.swift
import Foundation

struct BadiDate {
    let day: Int
    let monthName: String
    let year: Int

    var formatted: String {
        "\(day) \(monthName) \(year) BE"
    }
}

enum BadiCalendarConverter {
    private static let monthNames = [
        "Splendor",
        "Glory",
        "Beauty",
        "Grandeur",
        "Light",
        "Mercy",
        "Words",
        "Perfection",
        "Names",
        "Might",
        "Will",
        "Knowledge",
        "Power",
        "Speech",
        "Questions",
        "Honour",
        "Sovereignty",
        "Dominion",
        "The Days of Há",
        "Loftiness"
    ]

    static func date(from date: Date) -> BadiDate {
        let localCalendar = Calendar(identifier: .gregorian)
        let localComponents = localCalendar.dateComponents([.year, .month, .day], from: date)
        let gregorianYear = localComponents.year ?? 1844

        var utcCalendar = Calendar(identifier: .gregorian)
        utcCalendar.timeZone = TimeZone(secondsFromGMT: 0)!

        let utcDate = utcCalendar.date(from: DateComponents(
            year: localComponents.year,
            month: localComponents.month,
            day: localComponents.day
        )) ?? date

        var badiYear = gregorianYear - 1843
        var nawRuz = nawRuzDate(forGregorianYear: gregorianYear, calendar: utcCalendar)

        if utcDate < nawRuz {
            badiYear -= 1
            nawRuz = nawRuzDate(forGregorianYear: gregorianYear - 1, calendar: utcCalendar)
        }

        let daysSinceNawRuz = utcCalendar.dateComponents([.day], from: nawRuz, to: utcDate).day ?? 0
        let dayOfYear = daysSinceNawRuz + 1
        let intercalaryDays = isGregorianLeapYear(gregorianYear) ? 5 : 4

        if dayOfYear <= 18 * 19 {
            let monthNumber = Int(ceil(Double(dayOfYear) / 19.0))
            let day = ((dayOfYear - 1) % 19) + 1
            return BadiDate(day: day, monthName: monthNames[monthNumber - 1], year: badiYear)
        }

        if dayOfYear <= 18 * 19 + intercalaryDays {
            return BadiDate(day: dayOfYear - (18 * 19), monthName: monthNames[18], year: badiYear)
        }

        return BadiDate(
            day: dayOfYear - (18 * 19 + intercalaryDays),
            monthName: monthNames[19],
            year: badiYear
        )
    }

    private static func isGregorianLeapYear(_ year: Int) -> Bool {
        (year % 4 == 0 && year % 100 != 0) || (year % 400 == 0)
    }

    private static func nawRuzDate(forGregorianYear year: Int, calendar: Calendar) -> Date {
        calendar.date(from: DateComponents(year: year, month: 3, day: 21)) ?? Date()
    }
}
